File: scripts/migrate_sessions_to_turso.go
//go:build ignore

package main

import (
	"context"
	"database/sql"
	"encoding/binary"
	"fmt"
	"log"
	"math"
	"os"
	"strings"
	"time"

	_ "github.com/mattn/go-sqlite3"
	_ "github.com/tursodatabase/go-libsql"
)

func main() {
	tursoURL := os.Getenv("TURSO_DATABASE_URL")
	tursoToken := os.Getenv("TURSO_AUTH_TOKEN")
	localDB := os.Getenv("LOCAL_SESSIONS_DB")

	if tursoURL == "" || tursoToken == "" {
		log.Fatal("TURSO_DATABASE_URL and TURSO_AUTH_TOKEN must be set")
	}
	if localDB == "" {
		home, _ := os.UserHomeDir()
		localDB = home + "/.agentctl/storage/sessions.db"
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()

	// Open local SQLite database
	log.Printf("Opening local database: %s", localDB)
	local, err := sql.Open("sqlite3", localDB)
	if err != nil {
		log.Fatalf("Failed to open local database: %v", err)
	}
	defer local.Close()

	// Open Turso database
	dsn := fmt.Sprintf("%s?authToken=%s", tursoURL, tursoToken)
	log.Printf("Connecting to Turso: %s", tursoURL)
	turso, err := sql.Open("libsql", dsn)
	if err != nil {
		log.Fatalf("Failed to open Turso database: %v", err)
	}
	defer turso.Close()

	// Test Turso connection
	if err := turso.PingContext(ctx); err != nil {
		log.Fatalf("Failed to connect to Turso: %v", err)
	}
	log.Println("Connected to Turso successfully")

	// Query sessions with embeddings from local DB
	rows, err := local.QueryContext(ctx, `
		SELECT id, workspace_path, project_name, summary, accomplished, decisions,
		       gotchas, tags, key_files, started_at, ended_at, embedding,
		       embedding_model, created_at, updated_at
		FROM sessions
		WHERE embedding IS NOT NULL
	`)
	if err != nil {
		log.Fatalf("Failed to query local sessions: %v", err)
	}
	defer rows.Close()

	// Prepare insert statement for Turso
	insertSQL := `
		INSERT OR REPLACE INTO sessions
		(id, workspace_path, project_name, summary, accomplished, decisions,
		 gotchas, tags, key_files, started_at, ended_at, embedding,
		 embedding_model, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, vector(?), ?, ?, ?)
	`

	var migrated int
	for rows.Next() {
		var id, workspacePath, summary, accomplished, decisions, gotchas string
		var tags, keyFiles, startedAt, endedAt, createdAt, updatedAt string
		var projectName, embeddingModel sql.NullString
		var embeddingBlob []byte

		err := rows.Scan(&id, &workspacePath, &projectName, &summary, &accomplished,
			&decisions, &gotchas, &tags, &keyFiles, &startedAt, &endedAt,
			&embeddingBlob, &embeddingModel, &createdAt, &updatedAt)
		if err != nil {
			log.Printf("Failed to scan row: %v", err)
			continue
		}

		// Convert embedding blob to vector string format
		vectorStr := blobToVectorString(embeddingBlob)
		if vectorStr == "" {
			log.Printf("Skipping session %s: invalid embedding", id)
			continue
		}

		// Insert into Turso
		_, err = turso.ExecContext(ctx, insertSQL,
			id, workspacePath, nullStr(projectName), summary, accomplished,
			decisions, gotchas, tags, keyFiles, startedAt, endedAt,
			vectorStr, nullStr(embeddingModel), createdAt, updatedAt)
		if err != nil {
			log.Printf("Failed to insert session %s: %v", id, err)
			continue
		}

		migrated++
		log.Printf("Migrated session: %s (%.50s...)", id, summary)
	}

	if err := rows.Err(); err != nil {
		log.Fatalf("Error iterating rows: %v", err)
	}

	log.Printf("Migration complete: %d sessions migrated", migrated)

	// Verify by counting
	var count int
	err = turso.QueryRowContext(ctx, "SELECT COUNT(*) FROM sessions").Scan(&count)
	if err != nil {
		log.Printf("Failed to verify: %v", err)
	} else {
		log.Printf("Turso sessions table now has %d rows", count)
	}
}

func blobToVectorString(blob []byte) string {
	if len(blob) == 0 || len(blob)%4 != 0 {
		return ""
	}

	dims := len(blob) / 4
	values := make([]string, dims)

	for i := 0; i < dims; i++ {
		bits := binary.LittleEndian.Uint32(blob[i*4 : (i+1)*4])
		f := math.Float32frombits(bits)
		values[i] = fmt.Sprintf("%f", f)
	}

	return "[" + strings.Join(values, ",") + "]"
}

func nullStr(ns sql.NullString) any {
	if ns.Valid {
		return ns.String
	}
	return nil
}

File: scripts/migrate_sessions_to_turso.py
#!/usr/bin/env python3
"""Migrate sessions with embeddings from local SQLite to Turso."""

import os
import sqlite3
import struct
import libsql_experimental as libsql

def blob_to_vector_string(blob):
    """Convert embedding blob (float32 array) to vector string format."""
    if not blob:
        return None

    # Each float32 is 4 bytes
    num_floats = len(blob) // 4
    floats = struct.unpack(f'<{num_floats}f', blob)

    # Format as vector string: [f1,f2,f3,...]
    return '[' + ','.join(f'{f:.6f}' for f in floats) + ']'

def main():
    turso_url = os.environ.get('TURSO_DATABASE_URL')
    turso_token = os.environ.get('TURSO_AUTH_TOKEN')
    local_db = os.environ.get('LOCAL_SESSIONS_DB',
                              os.path.expanduser('~/.agentctl/storage/sessions.db'))

    if not turso_url or not turso_token:
        print("Error: TURSO_DATABASE_URL and TURSO_AUTH_TOKEN must be set")
        return 1

    print(f"Opening local database: {local_db}")
    local = sqlite3.connect(local_db)
    local.row_factory = sqlite3.Row

    print(f"Connecting to Turso: {turso_url}")
    turso = libsql.connect(database=turso_url, auth_token=turso_token)

    # Query sessions with embeddings
    cursor = local.execute("""
        SELECT id, workspace_path, project_name, summary, accomplished, decisions,
               gotchas, tags, key_files, started_at, ended_at, embedding,
               embedding_model, created_at, updated_at
        FROM sessions
        WHERE embedding IS NOT NULL
    """)

    migrated = 0
    for row in cursor:
        session_id = row['id']
        embedding_blob = row['embedding']
        vector_str = blob_to_vector_string(embedding_blob)

        if not vector_str:
            print(f"Skipping session {session_id}: invalid embedding")
            continue

        try:
            # Insert using vector() function for the embedding
            turso.execute("""
                INSERT OR REPLACE INTO sessions
                (id, workspace_path, project_name, summary, accomplished, decisions,
                 gotchas, tags, key_files, started_at, ended_at, embedding,
                 embedding_model, created_at, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, vector(?), ?, ?, ?)
            """, (
                row['id'],
                row['workspace_path'],
                row['project_name'],
                row['summary'],
                row['accomplished'],
                row['decisions'],
                row['gotchas'],
                row['tags'],
                row['key_files'],
                row['started_at'],
                row['ended_at'],
                vector_str,
                row['embedding_model'],
                row['created_at'],
                row['updated_at']
            ))
            migrated += 1
            summary_preview = (row['summary'] or '')[:50]
            print(f"Migrated: {session_id} ({summary_preview}...)")
        except Exception as e:
            print(f"Failed to insert session {session_id}: {e}")

    turso.commit()

    # Verify
    result = turso.execute("SELECT COUNT(*) FROM sessions").fetchone()
    print(f"\nMigration complete: {migrated} sessions migrated")
    print(f"Turso sessions table now has {result[0]} rows")

    local.close()
    turso.close()
    return 0

if __name__ == '__main__':
    exit(main())
